File: Data Structures & Algorithms/three-integer-sum/submission-1.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i-1]:
                continue

            left, right = i + 1, len(nums) - 1
            while left < right:
                if left > i + 1 and nums[left] == nums[left - 1]:
                    left += 1
                    continue

                if nums[i] + nums[left] + nums[right] == 0:
                    result.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1
                elif nums[i] + nums[left] + nums[right] > 0:
                    right -= 1
                else:
                    left += 1
        
        return result
        

File: Data Structures & Algorithms/trapping-rain-water/submission-1.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) {
        let leftMax = [], rightMax = []
        let left = height[0], right = height[height.length - 1]
        let result = 0
        let n = height.length

        for (let i = 0; i < n; i++) {
            leftMax.push(left)
            if (height[i] > left) {
                left = height[i]
            }
        }

        for (let i = n - 1; i >= 0; i--) {
            rightMax.push(right)
            if (height[i] > right) {
                right = height[i]
            }
        }

        for (let i = 0; i < n; i++) {
            let waterVolume = Math.min(leftMax[i], rightMax[n - i - 1])
            if (waterVolume > height[i]) {
                result += waterVolume - height[i]
            }
        }

        return result
    }
}
